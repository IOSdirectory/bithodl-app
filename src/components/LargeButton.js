import React, { Component } from 'react';
import { StyleSheet, TouchableOpacity, View, Dimensions, ActivityIndicator } from 'react-native';
import PropTypes from 'prop-types';
import { ifIphoneX } from 'react-native-iphone-x-helper';
import StyledText from './StyledText';

const windowDimensions = Dimensions.get('window');
const FULL_WIDTH = windowDimensions.width;
const DEFAULT_WIDTH = FULL_WIDTH - 80;

const styles = StyleSheet.create({
  button: {
    backgroundColor: '#FFCA8B',
    width: DEFAULT_WIDTH,
    height: 42,
    alignItems: 'center',
    justifyContent: 'center',
    borderRadius: 4,
    shadowColor: '#000000',
    shadowOpacity: 0.1,
    shadowOffset: { width: 0, height: 0 },
    shadowRadius: 4
  },
  fullWidth: {
    width: FULL_WIDTH,
    borderRadius: 0,
    bottom: ifIphoneX(-24, 0)
  },
  label: {
    fontWeight: '600',
    fontSize: 12,
    letterSpacing: 1,
    color: '#D59247'
  },
  disabled: {
    opacity: 0.4
  },
  loader: {
    height: 12,
    position: 'absolute',
    top: 15,
    left: 15
  }
});

export default class LargeButton extends Component {
  constructor(props) {
    super(...arguments);

    this.state = {
      disabled: props.disabled || false,
      loading: false,
      fullWidth: false
    };
  }

  componentDidMount() {
    /*
     * HACK: The promise from onPress might resolve and try to update state
     * after this component has been unmounted which leads to a warning.
     * One solution is to wrap the promise but this "hack" is easier in
     * this case.
     * <https://reactjs.org/blog/2015/12/16/ismounted-antipattern.html>
     */
    this._isMounted = true;
  }

  componentWillReceiveProps(nextProps) {
    if (this.props.disabled !== nextProps.disabled) {
      this.setState({
        disabled: nextProps.disabled
      });
    }

    if (this.props.fullWidth !== nextProps.fullWidth) {
      this.setState({
        fullWidth: nextProps.fullWidth
      });
    }
  }

  componentWillUnmount() {
    this._isMounted = false;
  }

  _onPress() {
    const onPress = this.props.onPress;
    const promise = onPress ? onPress() : null;

    if (promise instanceof Promise === false) {
      return;
    }

    this.setState({
      disabled: true,
      loading: true
    });

    promise.then(() => {
      if (!this._isMounted) {
        return;
      }

      this.setState({
        disabled: false,
        loading: false
      });
    });
  }

  render() {
    const { disabled, loading } = this.state;
    const label = loading ? this.props.loadingLabel : this.props.label;

    const buttonStyles = [
      styles.button,
      this.props.style
    ];

    if (disabled) {
      buttonStyles.push(styles.disabled);
    }

    if (this.state.fullWidth) {
      buttonStyles.push(styles.fullWidth);
    }

    return (
      <TouchableOpacity disabled={disabled} activeOpacity={0.85} onPress={this._onPress.bind(this)}>
        <View style={buttonStyles}>
          <ActivityIndicator animating={loading} color='#D59247' style={styles.loader} size='small' />
          <StyledText style={styles.label}>
            {label.toUpperCase()}
          </StyledText>
        </View>
      </TouchableOpacity>
    );
  }
}

LargeButton.propTypes = {
  label: PropTypes.string.isRequired,
  loadingLabel: PropTypes.string,
  onPress: PropTypes.func,
  style: PropTypes.any,
  fullWidth: PropTypes.bool,
  disabled: PropTypes.bool
};
