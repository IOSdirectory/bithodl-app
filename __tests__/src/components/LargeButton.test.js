import React from 'react';
import renderer from 'react-test-renderer';
import LargeButton from '../../../src/components/LargeButton';

describe('LargeButton', () => {
  it('renders correctly', () => {
    const tree = renderer.create(
      <LargeButton label='06b5f369-ec66-4e0a-8777-90c71dec3f7b' />
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('renders correctly in full width', () => {
    const tree = renderer.create(
      <LargeButton label='099349fd-1c84-415b-818f-c8c6af2ea8e7' fullWidth={true} />
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('renders correctly as disabled', () => {
    const tree = renderer.create(
      <LargeButton label='15013b10-4256-4d32-b25f-a031dfb6ce95' disabled={true} />
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });
});
