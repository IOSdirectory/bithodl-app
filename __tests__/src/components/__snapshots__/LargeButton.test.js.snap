// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LargeButton renders correctly 1`] = `
<View
  accessibilityComponentType={undefined}
  accessibilityLabel={undefined}
  accessibilityTraits={undefined}
  accessible={true}
  collapsable={undefined}
  hitSlop={undefined}
  isTVSelectable={true}
  nativeID={undefined}
  onLayout={undefined}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "opacity": 1,
    }
  }
  testID={undefined}
  tvParallaxProperties={undefined}
>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "backgroundColor": "#FFCA8B",
          "borderRadius": 4,
          "height": 42,
          "justifyContent": "center",
          "shadowColor": "#000000",
          "shadowOffset": Object {
            "height": 0,
            "width": 0,
          },
          "shadowOpacity": 0.1,
          "shadowRadius": 4,
          "width": 670,
        },
        undefined,
      ]
    }
  >
    <ActivityIndicator
      animating={false}
      color="#D59247"
      hidesWhenStopped={true}
      size="small"
      style={
        Object {
          "height": 12,
          "left": 15,
          "position": "absolute",
          "top": 15,
        }
      }
    />
    <Text
      accessible={true}
      allowFontScaling={true}
      disabled={false}
      ellipsizeMode="tail"
      numberOfLines={undefined}
      style={
        Array [
          Object {
            "color": "#504F63",
            "fontFamily": "Avenir Next",
            "fontWeight": "400",
          },
          Object {
            "color": "#D59247",
            "fontSize": 12,
            "fontWeight": "600",
            "letterSpacing": 1,
          },
        ]
      }
    >
      06B5F369-EC66-4E0A-8777-90C71DEC3F7B
    </Text>
  </View>
</View>
`;

exports[`LargeButton renders correctly as disabled 1`] = `
<View
  accessibilityComponentType={undefined}
  accessibilityLabel={undefined}
  accessibilityTraits={undefined}
  accessible={true}
  collapsable={undefined}
  hitSlop={undefined}
  isTVSelectable={true}
  nativeID={undefined}
  onLayout={undefined}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "opacity": 1,
    }
  }
  testID={undefined}
  tvParallaxProperties={undefined}
>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "backgroundColor": "#FFCA8B",
          "borderRadius": 4,
          "height": 42,
          "justifyContent": "center",
          "shadowColor": "#000000",
          "shadowOffset": Object {
            "height": 0,
            "width": 0,
          },
          "shadowOpacity": 0.1,
          "shadowRadius": 4,
          "width": 670,
        },
        undefined,
        Object {
          "opacity": 0.4,
        },
      ]
    }
  >
    <ActivityIndicator
      animating={false}
      color="#D59247"
      hidesWhenStopped={true}
      size="small"
      style={
        Object {
          "height": 12,
          "left": 15,
          "position": "absolute",
          "top": 15,
        }
      }
    />
    <Text
      accessible={true}
      allowFontScaling={true}
      disabled={false}
      ellipsizeMode="tail"
      numberOfLines={undefined}
      style={
        Array [
          Object {
            "color": "#504F63",
            "fontFamily": "Avenir Next",
            "fontWeight": "400",
          },
          Object {
            "color": "#D59247",
            "fontSize": 12,
            "fontWeight": "600",
            "letterSpacing": 1,
          },
        ]
      }
    >
      15013B10-4256-4D32-B25F-A031DFB6CE95
    </Text>
  </View>
</View>
`;

exports[`LargeButton renders correctly in full width 1`] = `
<View
  accessibilityComponentType={undefined}
  accessibilityLabel={undefined}
  accessibilityTraits={undefined}
  accessible={true}
  collapsable={undefined}
  hitSlop={undefined}
  isTVSelectable={true}
  nativeID={undefined}
  onLayout={undefined}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "opacity": 1,
    }
  }
  testID={undefined}
  tvParallaxProperties={undefined}
>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "backgroundColor": "#FFCA8B",
          "borderRadius": 4,
          "height": 42,
          "justifyContent": "center",
          "shadowColor": "#000000",
          "shadowOffset": Object {
            "height": 0,
            "width": 0,
          },
          "shadowOpacity": 0.1,
          "shadowRadius": 4,
          "width": 670,
        },
        undefined,
      ]
    }
  >
    <ActivityIndicator
      animating={false}
      color="#D59247"
      hidesWhenStopped={true}
      size="small"
      style={
        Object {
          "height": 12,
          "left": 15,
          "position": "absolute",
          "top": 15,
        }
      }
    />
    <Text
      accessible={true}
      allowFontScaling={true}
      disabled={false}
      ellipsizeMode="tail"
      numberOfLines={undefined}
      style={
        Array [
          Object {
            "color": "#504F63",
            "fontFamily": "Avenir Next",
            "fontWeight": "400",
          },
          Object {
            "color": "#D59247",
            "fontSize": 12,
            "fontWeight": "600",
            "letterSpacing": 1,
          },
        ]
      }
    >
      099349FD-1C84-415B-818F-C8C6AF2EA8E7
    </Text>
  </View>
</View>
`;
